### PR TITLE
add prometheus pushgateway service; adjust prometheus config

### PR DESCRIPTION
**Pull request by suiguoxin**
- add prometheus pushgateway service
- set `hornor_label` as true in Prometheus to keep `instance` label valuable

### CODE REVIEW

File: src/prometheus-pushgateway/deploy/service.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+cluster-type:
+  - yarn
+  - k8s
+
+prerequisite:
+  - cluster-configuration
+  - docker-cache
```
Comment (Binyang2014) [inline src/prometheus-pushgateway/deploy/service.yaml:24]: Do we need this?

File: src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    metadata:
+      name: prometheus-pushgateway
+      labels:
+        app: prometheus-pushgateway
+      annotations:
+        prometheus.io/scrape: "true"
+        prometheus.io/path: "/prometheus-pushgateway/metrics"
+        prometheus.io/port: "{{ prometheus_pushgateway_port }}"
+    spec:
+      hostNetwork: true
```
Comment (Binyang2014) [inline src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template:45]: Can we change hostNetwork to false?

File: src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      containers:
+      - name: prometheus-pushgateway
+        image: {{ cluster_cfg["cluster"]["docker-registry"]["prefix"] }}prometheus-pushgateway:{{ cluster_cfg["cluster"]["docker-registry"]["tag"] }}
+        imagePullPolicy: Always
+        {%- if cluster_cfg['cluster']['common']['qos-switch'] == "true" %}
+        resources:
+          limits:
+            memory: "10Gi"
```
Comment (Binyang2014) [inline src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template:53]: Why need 10Gi memory...

File: src/prometheus-pushgateway/deploy/service.yaml (hunk 1 above)
Comment (suiguoxin) [reply src/prometheus-pushgateway/deploy/service.yaml:24]: > Do we need this?

Yes, why not ?

File: src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template (hunk 3 above)
Comment (suiguoxin) [reply src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template]: > Why need 10Gi memory...

Changed to 256Mi limit

File: src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template (hunk 2 above)
Comment (suiguoxin) [reply src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template:45]: > Can we change hostNetwork to false?

Not sure, why change it?

File: src/prometheus-pushgateway/deploy/service.yaml (hunk 1 above)
Comment (Binyang2014) [reply src/prometheus-pushgateway/deploy/service.yaml:24]: I think docker-cache is an optional services.

File: src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template (hunk 2 above)
Comment (Binyang2014) [reply src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template:45]: using hostNetwork break the network isolation, If we don't need to access all network interface, set hostNetwork to false will make pod more security

Comment (suiguoxin) [reply src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template]: Changed

Comment (suiguoxin) [reply src/prometheus-pushgateway/deploy/prometheus-pushgateway.yaml.template]: We may remove this after pylon gets configured. Currently we'll not be able to access the Pushgateway portal with hostNetwork set as true.

**Review by Binyang2014 — APPROVED** [review]

**Pull request merged**